### PR TITLE
RBD Plugin: Pass monitors addresses in a comma-separed list instead of trying one by one.

### PR DESCRIPTION
**Pull request by cofyc**
**What this PR does / why we need it**:

In production, monitors may crash (or have a network problem), if we try monitors one by one, rbd
command will hang a long time (e.g. `rbd map -m <unconnectable_host_ip>`
on linux 4.4 timed out in 6 minutes) when trying a unconnectable monitor. This is unacceptable.

Actually, we can simply pass a comma-separated list monitor addresses to `rbd`
command utility. Kernel rbd/libceph modules will pick monitor randomly
and try one by one, `rbd` command utility succeed soon if there is a
good one in monitors list.

[Docs](http://docs.ceph.com/docs/jewel/man/8/rbd/#cmdoption-rbd-m) about `-m` option of `rbd` is wrong,  'rbd' utility simply pass '-m <mon>' parameter to kernel rbd/libceph modules, which
takes a comma-seprated list of one or more monitor addresses (e.g. ip1[:port1][,ip2[:port2]...]) in its first version in linux (see https://github.com/torvalds/linux/blob/602adf400201636e95c3fed9f31fba54a3d7e844/net/ceph/ceph_common.c#L239). Also, libceph choose monitor randomly, so we can simply pass all addresses without randomization (see https://github.com/torvalds/linux/blob/602adf400201636e95c3fed9f31fba54a3d7e844/net/ceph/mon_client.c#L132).

From what I saw, there is no need to iterate monitor hosts one by one.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

Run `rbd map` against unconnectable monitor address logs on Linux 4.4:

```
root@myhost:~# uname -a
Linux myhost 4.4.0-62-generic #83-Ubuntu SMP Wed Jan 18 14:10:15 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
root@myhost:~# time rbd map kubernetes-dynamic-pvc-941ff4d2-b951-11e7-8836-049fca8e58df --pool <pool> --id <id> -m <unconnectable_host_ip> --key=<password>
rbd: sysfs write failed
2017-12-20 18:55:11.810583 7f7ec56863c0  0 monclient(hunting): authenticate timed out after 300
2017-12-20 18:55:11.810638 7f7ec56863c0  0 librados: client.<id> authentication error (110) Connection timed out
rbd: couldn't connect to the cluster!
In some cases useful info is found in syslog - try "dmesg | tail" or so.
rbd: map failed: (110) Connection timed out

real	6m0.018s
user	0m0.052s
sys	0m0.064s
```  

We can simply pass a comma-separated list of monitors, if there is a good one in them, `rbd map` succeed soon.

```
root@myhost:~# time rbd map kubernetes-dynamic-pvc-941ff4d2-b951-11e7-8836-049fca8e58df --pool <pool> --id <id> -m <unconnectable_host_ip>,<good_host_ip> --key=<password>

/dev/rbd3

real	0m0.426s
user	0m0.008s
sys	0m0.008s
```

**Release note**:

```release-note
Pass monitors in comma-separated list in RBD plugin to improve availability.
```

### CODE REVIEW

**Comment by rootfs** [comment]
[(jewel) rbd docs](http://docs.ceph.com/docs/jewel/man/8/rbd/) doesn't seem to support coma separated mon list, but my rbd client does work in this way

adding @dillaman

**Comment by dillaman** [comment]
@rootfs The "-m" / "--mon_host" optional accepts comma, semi-colon, or white-space delimited list of monitor IP addresses; or a DNS name to resolve [1]

[1] http://docs.ceph.com/docs/master/rados/configuration/ceph-conf/#reading-configuration-metadata-at-runtime

**Comment by rootfs** [comment]
@dillaman thank you!

File: pkg/volume/rbd/rbd_util.go [hunk 1]
```diff
@@ -133,6 +132,17 @@ func rbdErrors(runErr, resultErr error) error {
 	return resultErr
 }
 
+// 'rbd' utility simply pass '-m <mon>' parameter to kernel rbd/libceph
```
Comment (rootfs) [inline pkg/volume/rbd/rbd_util.go]: passes

File: pkg/volume/rbd/rbd_util.go [hunk 2]
```diff
@@ -133,6 +132,17 @@ func rbdErrors(runErr, resultErr error) error {
...
+// modules, which takes a comma-seprated list of one or more monitor addresses
```
Comment (rootfs) [inline pkg/volume/rbd/rbd_util.go]: take

File: pkg/volume/rbd/rbd_util.go (hunk 1 above)
Comment (dillaman) [reply pkg/volume/rbd/rbd_util.go]: Nit: technically, the "-m" / "--mon_host" CLI optional is used for the CLI to connect to one or more monitors, grab the full monitor map, and it builds its own comma-separated list of IP addresses to send to the kernel.

File: pkg/volume/rbd/rbd_util.go [hunk 3]
```diff
@@ -156,88 +166,79 @@ func (util *RBDUtil) rbdLock(b rbdMounter, lock bool) error {
...
-			// clean up orphaned lock if no watcher on the image
-			used, rbdOutput, statusErr := util.rbdStatus(&b)
-			if statusErr != nil {
-				return fmt.Errorf("rbdStatus failed error %v, rbd output: %v", statusErr, rbdOutput)
-			}
-			if used {
-				// this image is already used by a node other than this node
-				return fmt.Errorf("rbd image: %s/%s is already used by a node other than this node, rbd output: %v", b.Image, b.Pool, output)
-			}
+	if lock {
```
Comment (rootfs) [inline pkg/volume/rbd/rbd_util.go]: since no locking is needed, can you clean it up too?

Comment (cofyc) [reply pkg/volume/rbd/rbd_util.go]: Removed in https://github.com/kubernetes/kubernetes/pull/57512/commits/4598f0df1ec74122218fd5d7082930f9264d5e83.

File: pkg/volume/rbd/rbd_util.go (hunk 1 above)
Comment (cofyc) [reply pkg/volume/rbd/rbd_util.go]: @rootfs Comments are updated in https://github.com/kubernetes/kubernetes/pull/57512/commits/4598f0df1ec74122218fd5d7082930f9264d5e83#diff-36f18e327f36d95eb1333c4a18781184L135.

**Comment by cofyc** [comment]
@rootfs Comments are fixed and unnecessary code is removed in https://github.com/kubernetes/kubernetes/pull/57512/commits/b196301b67195b66fb28a12feef767ab86383a02.

**Comment by rootfs** [comment]
thanks!

**Comment by cofyc** [comment]
```release-note
Pass monitors in comma-separated list in RBD plugin to improve availability.
```

**Pull request merged**